src/components/floating-menu/floating-menu.js

```diff
@@ -5,6 +5,72 @@ import trackBlur from '../../globals/js/mixins/track-blur';
 import getLaunchingDetails from '../../globals/js/misc/get-launching-details';
 import optimizedResize from '../../globals/js/misc/resize';
 
+/**
+ * The structure for the position of floating menu.
+ * @typedef {Object} FloatingMenu~position
+ * @property {number} left The left position.
+ * @property {number} top The top position.
+ * @property {number} right The right position.
+ * @property {number} bottom The bottom position.
+ */
+
+/**
+ * The structure for the size of floating menu.
+ * @typedef {Object} FloatingMenu~size
+ * @property {number} width The width.
+ * @property {number} height The height.
+ */
+
+/**
+ * The structure for the position offset of floating menu.
+ * @typedef {Object} FloatingMenu~offset
+ * @property {number} top The top position.
+ * @property {number} left The left position.
+ */
+
+export const DIRECTION_LEFT = 'left';
+export const DIRECTION_TOP = 'top';
+export const DIRECTION_RIGHT = 'right';
+export const DIRECTION_BOTTOM = 'bottom';
+
+/**
+ * @param {Object} params The parameters.
+ * @param {FloatingMenu~size} params.menuSize The size of the menu.
+ * @param {FloatingMenu~position} params.refPosition The position of the triggering element.
+ * @param {FloatingMenu~offset} [params.offset={ left: 0, top: 0 }] The position offset of the menu.
+ * @param {string} [params.direction=bottom] The menu direction.
+ * @param {number} [params.scrollY=0] The scroll position of the viewport.
+ * @returns {FloatingMenu~offset} The position of the menu, relative to the top-left corner of the viewport.
+ * @private
+ */
+export const getFloatingPosition = ({ menuSize, refPosition, offset = {}, direction = DIRECTION_BOTTOM, scrollY = 0 }) => {
+  const { left: refLeft = 0, top: refTop = 0, right: refRight = 0, bottom: refBottom = 0 } = refPosition;
+
+  const { width, height } = menuSize;
+  const { top = 0, left = 0 } = offset;
+  const refCenterHorizontal = (refLeft + refRight) / 2;
+  const refCenterVertical = (refTop + refBottom) / 2;
+
+  return {
+    [DIRECTION_LEFT]: () => ({
```

Comment:
Do these need to be methods? Or can the return just be:

```
return {
  [DIRECTION_LEFT]: {
    // ...
  },
  // ...
}[direction];

Reply:
HI @joshblack thank you for your feedback. I used the technique of [thunk](https://github.com/gaearon/redux-thunk#whats-a-thunk) to avoid evaluating expressions for all directions when we are only handling one direction. But I don’t think it’ll be a big deal and I can definitely make a change if you feel strongly about pure object here.

Reply:
Thunks make sense 😄 didn't realize the performance path was so critical here. Is it more expensive to do the math than the closure allocations? Especially given the arrow function bindings

Reply:
Perfect point - I don't feel big one way vs. other, so please don't hesitate to let me know you'd like non-thunk path and I can definitely make a change. Thanks!

Reply:
If this is for optimizing against performance, I would theorize that the thunk allocation here trumps the cost of performing the numeric calculations (in this specific case). This is further compounded by the fact that `getFloatingPosition` triggers allocations of new closures for each direction during each invocation. Obviously no numbers here though 😞  just based off of theoretical guesses deriving from: https://www.reaktor.com/blog/javascript-performance-fundamentals-make-bluebird-fast/

Reply:
OK fair enough @joshblack - Went with non-thunk approach for now.